ports/stm32: Add support for the Arduino Opta board.

iabdalkader opened the pull request.

iabdalkader committed:
stm32/boards/ARDUINO_OPTA: Add support for the Arduino Opta board.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/board_init.c:
```diff
@@ -0,0 +1,94 @@
+}
+
+void OPTA_board_early_init(void) {
+    HAL_InitTick(0);
```
Can you use `TICK_INT_PRIORITY` here instead of 0?

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/board_init.c:
```diff
@@ -0,0 +1,94 @@
+    // This is enabled in the bootloader anyway.
+    OPTA_board_osc_enable(true);
+
+    #if MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE == 1
```
Suggest just `#if MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE`, for consistency with the rest of the code.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/board_init.c:
```diff
@@ -0,0 +1,94 @@
+void OPTA_board_enter_bootloader(void) {
+    RTC_HandleTypeDef RTCHandle;
+    RTCHandle.Instance = RTC;
+    HAL_RTCEx_BKUPWrite(&RTCHandle, RTC_BKP_DR0, 0xDF59);
```
Maybe this magic number can be a `#define` constant?

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/board_init.c:
```diff
@@ -0,0 +1,94 @@
+
+void OPTA_board_low_power(int mode) {
+
+    #if MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE == 0
```
Suggest `#if !MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE`.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/board_init.c:
```diff
@@ -0,0 +1,94 @@
+}
+
+void OPTA_board_osc_enable(int enable) {
+    mp_hal_pin_config(pyb_pin_OSCEN, MP_HAL_PIN_MODE_OUTPUT, MP_HAL_PIN_PULL_UP, 0);
```
Does it need the pull-up if it's output?

iabdalkader replied:
I did not experiment with this, but it doesn't have an external pull-up and all the other boards enable it.

dpgeorge replied:
Well, having a pull up/down with an output pin is not very useful, it won't really do anything except waste a tiny bit of power.  But we can leave this as-is if it's been tested and is working.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/deploy.md:
````diff
@@ -0,0 +1,18 @@
+dfu-util -w -a 0 -d 2341:0364 -D build-ARDUINO_OPTA/firmware.dfu
+```
+
+Or from MicroPython source repository, making sure the `cwd` is `ports/stm32`:
````
Maybe just write "current directory" instead of cwd.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/mpconfigboard.h:
```diff
@@ -0,0 +1,239 @@
+#define MICROPY_HW_ENABLE_RTC       (1)
+#define MICROPY_HW_ENABLE_RNG       (1)
+#define MICROPY_HW_ENABLE_ADC       (1)
+#define MICROPY_HW_ENABLE_DAC       (0)
```
You can leave this line out because disabled is the default.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/mpconfigboard.h:
```diff
@@ -0,0 +1,239 @@
+#define MICROPY_HW_ENABLE_SERVO     (0)
+#define MICROPY_HW_ENABLE_TIMER     (1)
+#define MICROPY_HW_ENABLE_SDCARD    (0)
+#define MICROPY_HW_ENABLE_MMCARD    (0)
```
You can leave SERVO, SDCARD and MMCARD out, because their default is disabled.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/mpconfigboard.h:
```diff
@@ -0,0 +1,239 @@
+
+void OPTA_board_osc_enable(int enable);
+#define MICROPY_BOARD_OSC_ENABLE    OPTA_board_osc_enable(1);
+#define MICROPY_BOARD_OSC_DISABLE   OPTA_board_osc_enable(0);
```
I don't think these two macros are used anywhere?

iabdalkader replied:
They're not currently used upstream but I use them in stop mode. I can send this patch too if you like, it will not have any effect on other boards:

```diff
diff --git a/ports/stm32/powerctrl.c b/ports/stm32/powerctrl.c
index 775514922..38e44faaf 100644
--- a/ports/stm32/powerctrl.c
+++ b/ports/stm32/powerctrl.c
@@ -823,6 +823,10 @@ void powerctrl_enter_stop_mode(void) {
     }
     #endif
 
+    #if defined(MICROPY_BOARD_OSC_DISABLE)
+    MICROPY_BOARD_OSC_DISABLE
+    #endif
+
     #if defined(STM32WB)
     powerctrl_low_power_prep_wb55();
     #endif
@@ -833,6 +837,10 @@ void powerctrl_enter_stop_mode(void) {
     HAL_PWR_EnterSTOPMode(PWR_LOWPOWERREGULATOR_ON, PWR_STOPENTRY_WFI);
     #endif
 
+    #if defined(MICROPY_BOARD_OSC_ENABLE)
+    MICROPY_BOARD_OSC_ENABLE
+    #endif
+
     // reconfigure the system clock after waking up
 
     #if defined(STM32F0)
```

dpgeorge replied:
That patch is useful.  But I'd suggest naming the macros something like `MICROPY_BOARD_PRE_STOP` and `MICROPY_BOARD_POST_STOP`.  That allows also having a `MICROPY_BOARD_PRE_STANDBY`.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/mpconfigboard.h:
```diff
@@ -0,0 +1,239 @@
+#endif
+
+// 4 wait states
+#define MICROPY_HW_FLASH_LATENCY    FLASH_LATENCY_2
```
If you want 4 wait states, this should be `FLASH_LATENCY_3`.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/mpconfigboard.h:
```diff
@@ -0,0 +1,239 @@
+#define MICROPY_HW_I2C3_SDA         (pyb_pin_I2C3_SDA)
+
+// USRSW is pulled low. Pressing the button makes the input go high.
+#define MICROPY_HW_USRSW_PIN        (pyb_pin_USER_BUTTON)
```
The comment here doesn't match the settings, which seem to indicate it's pulled high and pressing makes it go low.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/mpconfigboard.h:
```diff
@@ -0,0 +1,239 @@
+#define MICROPY_HW_ETH_RMII_TXD0    (pin_G13)
+#define MICROPY_HW_ETH_RMII_TXD1    (pin_G12)
+
+#define MICROPY_HW_USB_VID                      0x2341
```
Please move these settings up to the same block as `MICROPY_HW_USB_FS` is defined.

iabdalkader replied:
I moved this line down to better match the other Arduino board (they all have this split).

dpgeorge replied:
OK.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/pins.csv:
```diff
@@ -0,0 +1,95 @@
+-ETH_RMII_RXD1,-PC5
+-ETH_RMII_TX_EN,PG11
+-ETH_RMII_TXD0,PG13
+-ETH_RMII_TXD1,PG12
```
Please add a newline at the end of the file (not a blank line, a newline on the last line).

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/stm32h747.ld:
```diff
@@ -0,0 +1,51 @@
+
+/* Location of filesystem flash storage */
+_micropy_hw_internal_flash_storage_start = ORIGIN(FLASH_FS);
+_micropy_hw_internal_flash_storage_end = ORIGIN(FLASH_FS) + LENGTH(FLASH_FS);
```
Do you intend for this board to have an optional internal filesystem, that the user can enable manually?

iabdalkader replied:
Yes, they all do, if someone wants to customize the board, remove QSPI for example, there's a free sector that can be used for a small internal flash FS.

dpgeorge replied:
OK.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/wifi_nvram_1dx.h:
```diff
@@ -0,0 +1,49 @@
+static const uint8_t wifi_nvram_4343[] CYW43_RESOURCE_ATTRIBUTE =
```
Is there a reason you need a custom nvram file?  Or can you use the one supplied by cyw43-driver?  If this custom one is needed then maybe add a comment explaining why.

iabdalkader replied:
> Is there a reason you need a custom nvram file

Yes they all use a custom nvram file to disable the LPO.

dpgeorge replied:
Ah, I see.  Very good.

dpgeorge commented on ports/stm32/boards/ARDUINO_OPTA/stm32h7xx_hal_conf.h:
```diff
@@ -0,0 +1,51 @@
+#define DATA_CACHE_ENABLE           1
+#define INSTRUCTION_CACHE_ENABLE    1
+#define DATA_CACHE_ENABLE           1
+#define INSTRUCTION_CACHE_ENABLE    1
```
It looks like these two lines are duplicated (also in the other Arduino boards).

iabdalkader committed:
stm32/powerctrl: Add pre/post stop mode entry macros.

These allow a board to perform actions before/after stop mode.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards/ARDUINO_PORTENTA_H7: Update board config files.

Changes are:
- Add pre/post stop mode entry macros.
- Make SE05X pin accessible.
- Remove duplicate HAL macros.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards/ARDUINO_NICLA_VISION: Update board config files.

Changes are:
- Add pre/post stop mode entry macros.
- Make SE05X pin accessible.
- Remove duplicate HAL macros.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards/ARDUINO_GIGA: Update board config files.

To remove duplicate HAL macros.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge commented:
Thanks for updating, it now looks good, and I tested that all four stm32 Arduino boards build.
